### PR TITLE
fix(ci): handle disk mounting and logs reading edge-cases

### LINKED ISSUE #7564: devops: CI failure in `tee` when Rust tests pass
### Describe the issue or request

It seems this is a new issue in the CI. I am not sure how often it is happening, i only saw it once by now but it might worth to have a tciket in case it gets repeated.

A pull request failed with 

```
tee: 'standard output': Broken pipe
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 44 filtered out; finished in 2.11s
sudo docker exit status: 101
Error: Process completed with exit code 101.
```

at https://github.com/ZcashFoundation/zebra/actions/runs/6165092427/job/16733752210?pr=7515

Apparently, "The Rust test passes, but Docker returns an error status. This is probably an issue with the self-hosted runner changes." (that was mentioned by @teor2345 in slack).

### Expected Behavior

No failure or more details on what is the failure inside docker.

### Current Behavior

Failure with `sudo docker exit status: 101`

### Possible Solution

We need to figure out if this was a temporal issue, we might have to do nothing to resolve the ticket other than just a short comment of what went wrong, provide what to do if it happens again.

### Additional Information/Context

It looks like the failure is unrelated to what the pull request was doing.

### Is this happening on PRs?

yes

### Is this happening on the main branch?

not seen yet

### LINKED ISSUE #7614: bug: Timeout in Test testnet Zebra CD Docker config file job
### What happened?

I expected to see this happen:

CI passes. (Even though the job is not required.)

Instead, this happened: 

CI failed in https://github.com/ZcashFoundation/zebra/actions/runs/6280142047/job/17057280870?pr=7611 on PR #7611.

### What were you doing when the issue happened?

Marek opened PR #7611 with new checkpoints, which isn't related to this test at all.

### Zebra logs

The logs that are needed to diagnose this issue are in the CI workflow summary:
```
The job running on runner GitHub Actions 2 has exceeded the maximum execution time of 15 minutes.
```

https://github.com/ZcashFoundation/zebra/actions/runs/6280142047?pr=7611

These logs aren't enough to diagnose this bug by themselves.

```
2023-09-23T00:25:42.491019Z  INFO {zebrad="f510c3d" net="Test"}: zebrad::components::sync::end_of_support: Release always valid in Testnet
2023-09-23T00:25:51.264695Z  INFO {zebrad="f510c3d" net="Test"}:sync:try_to_sync:obtain_tips:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=400 current_range=(Excluded(Height(0)), Included(Height(400)))
2023-09-23T00:25:52.477096Z  INFO {zebrad="f510c3d" net="Test"}:peer_cache_updater: zebra_network::config: updated cached peer IP addresses cached_ip_count=2 peer_cache_file="/root/.cache/zebra/network/testnet.peers"
2023-09-23T00:25:53.719618Z  INFO {zebrad="f510c3d" net="Test"}:sync:try_to_sync: zebrad::components::sync: extending tips tips.len=1 in_flight=371 extra_hashes=0 lookahead_limit=1000 state_tip=Some(Height(400))
2023-09-23T00:26:02.803103Z  INFO {zebrad="f510c3d" net="Test"}:sync:try_to_sync:extend_tips:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=400 current_range=(Excluded(Height(400)), Included(Height(800)))
2023-09-23T00:26:11.504406Z  INFO {zebrad="f510c3d" net="Test"}:crawl_and_dial{new_peer_interval=61s}:crawl: zebra_network::peer_set::candidate_set: timeout waiting for peer service readiness or peer responses
2023-09-23T00:26:14.412064Z  INFO {zebrad="f510c3d" net="Test"}:sync:try_to_sync: zebrad::components::sync: extending tips tips.len=1 in_flight=741 extra_hashes=0 lookahead_limit=1000 state_tip=Some(Height(800))
2023-09-23T00:26:27.41[229](https://github.com/ZcashFoundation/zebra/actions/runs/6280142047/job/17057280870?pr=7611#step:4:230)8Z  INFO {zebrad="f510c3d" net="Test"}:sync:try_to_sync:extend_tips:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=400 current_range=(Excluded(Height(800)), Included(Height(1200)))
2023-09-23T00:26:32.504985Z  INFO {zebrad="f510c3d" net="Test"}: zebrad::components::sync::progress: estimated progress to chain tip sync_percent=0.052% current_height=Height(1200) network_upgrade=BeforeOverwinter remaining_sync_blocks=[231](https://github.com/ZcashFoundation/zebra/actions/runs/6280142047/job/17057280870?pr=7611#step:4:232)8722 time_since_last_state_block=0s
2023-09-23T00:26:32.504985Z  INFO {zebrad="f510c3d" net="Test"}: zebrad::components::sync::progress: estimated progress to chain tip sync_percent=0.052% current_height=Height(1200) network_upgrade=BeforeOverwinter remaining_sync_blocks=2318722 time_since_last_state_block=0s
tee: 'standard output': Broken pipe
Error: The operation was canceled.
```

### Zebra Version

zebrad 1.2.0 commit f510c3d

### Which operating systems does the issue happen on?

- [X] Linux
- [ ] macOS
- [ ] Windows
- [ ] Other OS

### OS details

CI

### Additional Information

We should increase the timeout on this job.

### LINKED ISSUE #7659: CI failing with mounting volume error
https://github.com/ZcashFoundation/zebra/actions/runs/6382508836/job/17321768540?pr=7653#step:13:188

```
docker: Error response from daemon: error while mounting volume '/var/lib/docker/volumes/fully-synced-rpc-85b855b/_data': failed to mount local volume: mount /dev/sdb:/var/lib/docker/volumes/fully-synced-rpc-85b855b/_data: device or resource busy.
Error: Process completed with exit code 125.
```

This seems to be happening for every PR we have opened *recently*. We can post more links to failures and more information in this ticket.

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

Some tests are still failing with `tee: 'standard output': Broken pipe`, recently it has been more frequently with the `sync-to-checkpoint` test, but it could happen on other tests too.

Also, mounting errors are still happening, and we'd like to capture the `dmesg` when this happens at `launch`, for better debugging. Other options like `set -x` will also help debugging and replicating some commands locally.

**Edit:**
In the end we had this different outputs for disks from the same VM:
```shell
NAME      MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
sda         8:0    0  300G  0 disk 
sdb         8:16   0   50G  0 disk 
|-sdb1      8:17   0 45.8G  0 part /var/lib/toolbox
|                                  /var/lib/google
|                                  /var/lib/docker
|                                  /var/lib/containerd
|                                  /var
|                                  /home
|                                  /mnt/stateful_partition
|-sdb2      8:18   0   16M  0 part 
|-sdb3      8:19   0    2G  0 part 
| `-vroot 253:0    0  1.9G  1 dm   /
|-sdb4      8:20   0   16M  0 part 
|-sdb5      8:21   0    2G  0 part 
|-sdb6      8:22   0  512B  0 part 
|-sdb7      8:23   0  512B  0 part 
|-sdb8      8:24   0   16M  0 part /usr/share/oem
|-sdb9      8:25   0  512B  0 part 
|-sdb10     8:26   0  512B  0 part 
|-sdb11     8:27   0    8M  0 part 
`-sdb12     8:28   0   32M  0 part 
```

```shell
NAME      MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
sda         8:0    0   50G  0 disk 
|-sda1      8:1    0 45.8G  0 part /var/lib/toolbox
|                                  /var/lib/google
|                                  /var/lib/docker
|                                  /var/lib/containerd
|                                  /var
|                                  /home
|                                  /mnt/stateful_partition
|-sda2      8:2    0   16M  0 part 
|-sda3      8:3    0    2G  0 part 
| `-vroot 253:0    0  1.9G  1 dm   /
|-sda4      8:4    0   16M  0 part 
|-sda5      8:5    0    2G  0 part 
|-sda6      8:6    0  512B  0 part 
|-sda7      8:7    0  512B  0 part 
|-sda8      8:8    0   16M  0 part /usr/share/oem
|-sda9      8:9    0  512B  0 part 
|-sda10     8:10   0  512B  0 part 
|-sda11     8:11   0    8M  0 part 
`-sda12     8:12   0   32M  0 part 
sdb         8:16   0  300G  0 disk 
```

Based on the preceding output, we can see that GCP is mounting the disks in **different order** on each VM. This is why the `/dev/sda` and `/dev/sdb` devices are not consistent across the VMs, nor across reboots on the same VM.

Fixes #7564
Fixes #7659
Fixes #7614

## Solution

For `tee: 'standard output': Broken pipe`:
- Use `shell: /usr/bin/bash -exo pipefail {0}` in GitHub jobs, for tests deployed to GCP
- Temporarily disable "set -e" to handle the broken pipe error gracefully
- Handle the exit code a bit better when reading from `grep` and `docker wait`
- Use `trap` quotes correctly
- Standardized `--command` quoting for GCP tests

For `failed to mount local volume: mount /dev/sdb device or resource busy`:

To work around this we're extracting the disk to be use, based on the device name we set in previous GitHub Action steps.
```shell
# Extract the correct disk name based on the device-name
DISK_NAME=$(ls -l /dev/disk/by-id | grep -oE "google-${{ inputs.test_id }}-${{ env.GITHUB_SHA_SHORT }} -> ../../[^ ]+" | grep -oE "/[^/]+$" | cut -c 2-); \
echo "Disk name: $DISK_NAME";
```

## Review
- Confirm this is working with `sync-to-checkpoint` test.
- Make the tests fail and confirm we're getting the required EXIT_STATUS code.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

**Comment by teor2345** [comment]
The sync-to-checkpoint test won't run any more because we have a checkpoint disk now. So it won't run on PRs or in the scheduled jobs.

I think the `tee` error might be caused by `set -o pipefail`, or the input or output processes exiting early. Either way, replacing `tee…` with `(tee … || true)` should fix it. (That runs `tee` in a subshell and ignores its exit status.)

There might also be arguments we can pass to `tee` to change its exit status, like `--mode=warn-nopipe`:
https://www.man7.org/linux/man-pages/man1/tee.1.html

**Comment by gustavovalverde** [comment]
@teor2345 yeah, at least I was able to capture the errors before the checkpoint disk got fully build and the job started to get skipped.

**Comment by teor2345** [comment]
> @teor2345 yeah, at least I was able to capture the errors before the checkpoint disk got fully build and the job started to get skipped.

Let's see if the error happens on other jobs? If it doesn't then maybe we can lower the priority of this PR and ticket.

**Comment by gustavovalverde** [comment]
It's happening here too: https://github.com/ZcashFoundation/zebra/actions/runs/6450509996/job/17510016301#step:4:240

**Comment by teor2345** [comment]
It might be worth trying `tee --mode=warn-nopipe ...` or `(tee … || true)` if you're looking for a quick fix.

**Comment by teor2345** [comment]
> It might be worth trying `tee --mode=warn-nopipe ...` 

It looks like `--mode=exit-nopipe` doesn't work, because we need `tee` to exit when `grep` finds something, and it's not doing that here:
https://github.com/ZcashFoundation/zebra/actions/runs/6450645551/job/17510271960?pr=7690#step:4:291

That means all the other modes won't work either.

But `(tee … || true)` will definitely ignore the exit status.

**Comment by gustavovalverde** [comment]
This should be good to go

**Review by teor2345 — DISMISSED** [review]
Thanks, that's a lot better.

**Comment by teor2345** [comment]
Unfortunately this PR failed in the merge queue due to issue #7659, so it might help to focus on fixing that issue first.

**Comment by gustavovalverde** [comment]
Yeah, I'll now focus on #7659

**Comment by teor2345** [comment]
> Yeah, I'll now focus on #7659

I've tried another way to un-block that issue in PR #7703: repeating the failing job until it works. It's not a long-term solution, but it would let us merge other PRs while we're fixing it.

File: .github/workflows/deploy-gcp-tests.yml [hunk 1]
```diff
@@ -230,7 +228,10 @@ jobs:
           --ssh-flag="-o ConnectionAttempts=20" \
           --ssh-flag="-o ConnectTimeout=5" \
           --command=' \
-          sudo dmesg \
+          lsblk;
+          sudo lsof /dev/sdb;
```
Comment (arya2) [inline .github/workflows/deploy-gcp-tests.yml:232]: Should this also be `/dev/$DISK_NAME`?

**Review by arya2 — APPROVED** [review]
Looks good to me, sync-to-checkpoint test passed here: https://github.com/ZcashFoundation/zebra/actions/runs/6455786703/job/17524400741

**Comment by teor2345** [comment]
> Looks good to me, sync-to-checkpoint test passed here: [ZcashFoundation/zebra/actions/runs/6455786703/job/17524400741](https://github.com/ZcashFoundation/zebra/actions/runs/6455786703/job/17524400741)

That's the sync-past-checkpoint test, this PR also needs a full sync to verify it works. @gustavovalverde did you want to run a full sync now, or wait until Friday?

File: .github/workflows/deploy-gcp-tests.yml (hunk 1 above)
Comment (teor2345) [reply .github/workflows/deploy-gcp-tests.yml:232]: Yes, @gustavovalverde can we fix the hard-coded sdb in another PR?

**Comment by gustavovalverde** [comment]
@teor2345 last Friday it failed, so I triggered a new one here:

https://github.com/ZcashFoundation/zebra/actions/runs/6460865308

File: .github/workflows/deploy-gcp-tests.yml (hunk 1 above)
Comment (gustavovalverde) [reply .github/workflows/deploy-gcp-tests.yml:232]: Fixed at #7709

**Comment by gustavovalverde** [comment]
Running https://github.com/ZcashFoundation/zebra/actions/runs/6460865308 failed, as that's basically the only test not needing a cached state and thus the only one which was not tested in this PR 😞 

It should be fixed now in:
- #7709

I'm running a Full sync from that branch to confirm: https://github.com/ZcashFoundation/zebra/actions/runs/6461129131
